### DISCUSSION TITLE
Session is already started issue

### DISCUSSION THREAD

**Discussion by adrianbecker013** [php-webdriver/php-webdriver#1075] [General]
Hello,

We are facing following error:-
PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\SessionNotCreatedException: Session is already started in vendor/php-webdriver/webdriver/lib/Exception/WebDriverException.php:130

Could you please help us how to resolve it?

Thank you

**Comment by OndraM** [comment]
Hi, could you provide the code you use to start new session?

Also how do you start selenium server or other browser driver? (like chromedriver etc.) Or you use Laravel Dusk etc.?

**Reply by adrianbecker013** [reply]
Hello @OndraM,

Thanks for your reply!

We are starting by using "geckodriver" on command line :-
```
developer@25:php_scraper$ geckodriver
1695182747234   geckodriver     INFO    Listening on 127.0.0.1:4444
```


We are using following PHP code to submit a form and retrieve response from website

```
namespace Facebook\WebDriver;

use Facebook\WebDriver\Remote\DesiredCapabilities;
use Facebook\WebDriver\Remote\RemoteWebDriver;
use Facebook\WebDriver\Firefox\FirefoxOptions;

require_once('vendor/autoload.php');

$host = 'http://localhost:4444';

$desiredCapabilities = DesiredCapabilities::firefox();

// Disable accepting SSL certificates
$desiredCapabilities->setCapability('acceptSslCerts', false);

// Add arguments via FirefoxOptions to start headless firefox
$firefoxOptions = new FirefoxOptions();
$firefoxOptions->addArguments(['-headless']);
$desiredCapabilities->setCapability(FirefoxOptions::CAPABILITY, $firefoxOptions);

$driver = RemoteWebDriver::create($host, $desiredCapabilities);

try {
    // Navigate to a website
    $driver->get('https://example.com');

    // Find an element and interact with it
    $searchBox = $driver->findElement(WebDriverBy::name('q'));
    $searchBox->sendKeys('Selenium WebDriver');
    $searchBox->submit();

    // Wait for a few seconds to see the result
    sleep(5);

    // Print the page title
    echo 'Page title: ' . $webDriver->getTitle() . PHP_EOL;

} catch (Exception $e) {
    // WebDriver session does not exist
    echo "WebDriver session does not exist.";
}
// terminate the session and close the browser
$driver->quit();
```

Thank you:)

**Comment by OndraM** [comment]
This is because your `geckodriver` is already serving one session. This could happen eg. if you don't properly terminate the previous session (using `$driver->quit();`), then for some time the geckodriver will be "occupied" with the previous session until timeout occurs.

Geckodriver itself can only have one concurrent session. If you want another, either start another `geckodriver` on different port, or use Selenium server to do the dispatching and run more concurrent browser sessions. [See README.](https://github.com/php-webdriver/php-webdriver#c-selenium-standalone-server)

**Comment by adrianbecker013** [comment]
@OndraM, Thanks for your reply!

You mean that concurrent requests are not possible with Geckodriver right? and we should go with Selenium server.

I will check  Selenium server as per your suggestion.

Thank you!

**Comment by adrianbecker013** [comment]
@OndraM, As per your suggestion I have prepared demo for Selenium server

We are starting selenium server by using following command :-

```
$ java -jar selenium-server-4.12.1.jar hub
```

We are using following PHP code to submit a form and retrieve response from website.

```
namespace Facebook\WebDriver;

use Facebook\WebDriver\Remote\DesiredCapabilities;
use Facebook\WebDriver\Remote\RemoteWebDriver;
use Facebook\WebDriver\Firefox\FirefoxOptions;

require_once('vendor/autoload.php');

$host = 'http://localhost:4444/wd/hub';

$desiredCapabilities = DesiredCapabilities::firefox();

// Disable accepting SSL certificates
$desiredCapabilities->setCapability('acceptSslCerts', false);
$desiredCapabilities->setCapability('platform', 'linux');

// Add arguments via FirefoxOptions to start headless firefox
$firefoxOptions = new FirefoxOptions();
$firefoxOptions->addArguments(['-headless']);
$desiredCapabilities->setCapability(FirefoxOptions::CAPABILITY, $firefoxOptions);

$driver = RemoteWebDriver::create($host, $desiredCapabilities);

try {
    // Navigate to a website
    $driver->get('https://example.com');

    // Find an element and interact with it
    $searchBox = $driver->findElement(WebDriverBy::name('q'));
    $searchBox->sendKeys('Selenium WebDriver');
    $searchBox->submit();

    // Wait for a few seconds to see the result
    sleep(5);

    // Print the page title
    echo 'Page title: ' . $webDriver->getTitle() . PHP_EOL;

} catch (Exception $e) {
    // WebDriver session does not exist
    echo "WebDriver session does not exist.";
}
// terminate the session and close the browser
$driver->quit();
```

But we are facing following error:-

$ php selenium_standalone.php
PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\Internal\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"capabilities":{"firstMatch":[{"browserName":"firefox","platformName":"linux","moz:firefoxOptions":{"args":["-headless"],"prefs":{"reader.parse-on-load.enabled":false,"devtools.jsonview.enabled":false}},"acceptInsecureCerts":false}]},"desiredCapabilities":{"browserName":"firefox","platform":"linux","moz:firefoxOptions":{"args":["-headless"],"prefs":{"reader.parse-on-load.enabled":false,"devtools.jsonview.enabled":false}},"acceptSslCerts":false}}

Operation timed out after 180001 milliseconds with 0 bytes received in vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php:20
Stack trace:
#0 vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php(328): Facebook\WebDriver\Exception\Internal\WebDriverCurlException::forCurlError()
#1 vendor/php-webdriver/webdriver/lib/Remote/RemoteWebDriver.php(129): Facebook\WebDriver\Remote\HttpCommandExecutor->execute()
#2 selenium_standalone.php(27): Facebook\WebDriver\Remote\RemoteWebDriver::create()
#3 {main}
  thrown in vendor/php-webdriver/webdriver/lib/Exception/Internal/WebDriverCurlException.php on line 20

Thank you!

**Reply by OndraM** [reply]
Have a look at https://github.com/php-webdriver/php-webdriver/wiki/Selenium-server .

From quick look you are using wrong URL, use `$host = 'http://localhost:4444;`

If it won't start, look at the output of Selenium server, whether it is running and receiving requests or printing some error. It is possible it can't find Geckodriver or something like that.

**Comment by adrianbecker013** [comment]
@OndraM, Thank you for your reply!

I have changed $host = 'http://localhost:4444; as per your suggestion in our script and there is following output of Selenium server and still we are getting same error:-
18:31:50.386 WARN [SeleniumSpanExporter$1.lambda$export$3] - {"traceId": "bb191db2e1271c4ee22ae349209beb5a","eventTime": 1695646910383828913,"eventName": "HTTP request execution complete","attributes": {"http.flavor": 1,"http.handler_class": "org.openqa.selenium.grid.sessionqueue.local.LocalNewSessionQueue","http.host": "localhost:4444","http.method": "POST","http.request_content_length": "450","http.scheme": "HTTP","http.status_code": 500,"http.target": "\u002fsession"}}

**Reply by adrianbecker013** [reply]
@OndraM, Awaiting the update from you on this.
